rust/src/utils/ip_addr.rs

```diff
@@ -0,0 +1,184 @@
+/* Copyright (C) 2025 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+// Author: Jeff Lucovsky <jlucovsky@oisf.net>
+
+use std::ffi::{c_char, c_uchar};
+use std::net::Ipv6Addr;
+
+/// Writes the shortened IPv6 string into `out_buf`, which has length `out_len`.
+/// Returns the number of bytes written (excluding the null terminator),
+/// or 0 on error (invalid addr pointer, buffer too small, etc.)
+#[no_mangle]
+pub unsafe extern "C" fn SCIPv6Shorten(
+    addr: *const c_uchar, // pointer to 16-byte IPv6
+    out_buf: *mut c_char, // out buffer allocated by caller
+    out_len: usize,       // size of out buffer
+) -> usize {
+    if addr.is_null() || out_buf.is_null() {
+        return 0;
+    }
+
+    unsafe {
+        // get 16-byte IPv6 address
+        let bytes = std::slice::from_raw_parts(addr, 16);
+
+        // Convert &[u8] → Ipv6Addr
+        let ipv6 = match <&[u8; 16]>::try_from(bytes) {
```

Comment:
how might this fail? If we have 16 bytes we have a valid ipv6 address, right?

Reply:
As reported in https://github.com/OISF/suricata/pull/14433#discussion_r2603169563

I think you should use a `from` method that is MSRV-compatible